iroh-bytes/src/store/mem.rs

```diff
@@ -215,11 +216,11 @@ impl MapEntry<Store> for Entry {
         self.outboard.tree().size().0
     }
 
-    async fn outboard(&self) -> io::Result<PreOrderOutboard<MemFile>> {
+    async fn outboard(&self) -> io::Result<impl Outboard> {
         Ok(self.outboard.clone())
     }
 
-    async fn data_reader(&self) -> io::Result<MemFile> {
+    async fn data_reader(&self) -> io::Result<impl AsyncSliceReader> {
```

Comment:
this makes it impossible to store this in a struct without boxing it are we okay with that?

Reply:
> this makes it impossible to store this in a struct without boxing it are we okay with that?

I think it is fine.

Once you have impl traits you can't really do manual state machines in a nice way anymore, so the way to "store this in a struct" is to have an async block and let rustc make a struct for you. But that is the style you end up doing anyway if you don't want to go crazy.

You can still have functions, including async fns, that take an unnamed type. And you can put your logic into an async block and spawn it, which basically boxes it anyway.

I have tried the explicit associated type thing now for a while, and is just so painful to compose them. E.g. just a thin wrapper to add logging or any kind of logic requires either pages of boilerplate or... boxing. So 🤷...